# Part 1

In [1]:
from datascience import *
import numpy as np
from tqdm import tqdm
import json
from IPython.display import YouTubeVideo, HTML, display
from ipywidgets import interact, widgets
%matplotlib inline
%pip install openai # You may see an Error pop up, that's fine.


  Using cached openai-1.99.9-py3-none-any.whl.metadata (29 kB)
  Using cached jiter-0.10.0-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (5.2 kB)
Using cached openai-1.99.9-py3-none-any.whl (786 kB)
Using cached jiter-0.10.0-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (352 kB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2/2 [openai]2m1/2 [openai]
Note: you may need to restart the kernel to use updated packages.


In [2]:
"""
When writing code that will go to production (e.g. be seen by other people), 
you'll usually set your API key as an environment variable via your terminal.
"""
# Find the API key we've given you in EdStem
API_KEY = "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"

In [3]:
from openai import OpenAI
client = OpenAI(api_key = API_KEY)

In [4]:
dataset = Table.read_table("SBFv2.trn.csv")
np.random.seed(50)
theData = dataset.sample(k=1000, with_replacement = False)
theData


whoTarget | intentYN | sexYN | sexReason       | offensiveYN | annotatorGender | annotatorMinority | sexPhrase         | speakerMinorityYN | WorkerId             | HITId                          | annotatorPolitics | annotatorRace | annotatorAge | post                                                         | targetMinority                     | targetCategory | targetStereotype
nan       | 0.66     | 0.5   | oral sex?       | nan         | woman           | nan               | eat the cookies   | nan               | -8935932304856669427 | 3Z33IC0JC0GO2QZ7DPV62JGZ3EZV9N | liberal           | white         | 45           | Why do black people call black women oreos? Cause they e ... | nan                                | nan            | nan
nan       | 0        | 0     | nan             | 0           | man             | nan               | nan               | nan               | 1781840716988750702  | 3V7ICJJAZAA8W6VPRD8FKZBCG81B4R | liberal           | white         | 30           | BUT I SOON REALIZED THAT EFFICIENCY AND QUALITY NEED NOT ... | nan                                | nan            | nan
0         | 0.66     | 0     | nan             | 1           | man             | nan               | nan               | nan               | -2877140746380073212 | 3S4TINXCC0HC0Z57AXNOYXHHXZ4BO4 | cons              | white         | 42           | that 400bil in savings means less labor cost. less peopl ... | nan                                | nan            | nan
1         | 1        | 0     | nan             | 1           | woman           | nan               | nan               | 0                 | 7912096326098817047  | 31MBOZ6PAOLN4JUGB53RH0J5IXUCLO | liberal           | white         | 32           | A child climbs up an apple tree with a rope. When it lat ... | folks with mental illness/disorder | disabled       | nan
1         | 1        | 0     | nan             | 1           | woman           | none              | nan               | 0                 | -450836770736143243  | 338GLSUI435HB89GMGQT1SI3AIWSFQ | mod-liberal       | white         | 55           | This gay kill a young bulgarian man to protect some crim ... | black folks                        | race           | racist against skin color
1         | 1        | 1     | ejaculation     | 1           | man             | nan               | comes on his face | 0                 | -2877140746380073212 | 3UUSLRKAULXYUQBS77XSI1JM8JV7DG | cons              | white         | 42           | What’s the difference between a priest and acne? Acne wa ... | christian folks                    | culture        | all are child molesters
nan       | 0        | 0     | nan             | 0           | woman           | nan               | nan               | nan               | -5471608738051449110 | 3MZ3TAMYTLHPKKBT8ZW9EFQZ3C8RI2 | other             | white         | 35           | tomorrow, 3 fucking PM EST yall are paying for my fallen ... | nan                                | nan            | nan
nan       | 0        | 0     | nan             | 0           | woman           | nan               | nan               | nan               | 955703508749055793   | 3566S7OX5DD42VVGREDXHZIDMVZ71I | liberal           | asian         | 23           | TIL that a bunch of Retards think Men's Right's is actua ... | nan                                | nan            | nan
0         | 1        | 0     | nan             | 1           | woman           | nan               | nan               | nan               | 1993680211374950203  | 3XUSYT70ITVKCFB8M8CJ3M9Z0130DF | mod-liberal       | white         | 47           | What was Stephen Hawking’s last meal? His shoulder           | nan                                | nan            | nan
1         | 0.66     | 1     | referencing sex | 1           | woman           | none              | I fucked her      | 0                 | -3114316428557054065 | 37PGLWGSJT03XGYYIFNCD34L64ZIKN | mod-cons          | white         | 33           | Never m

The social bias interference corpus is a model created to determine how and why biased and offensive text is deemed bias. The dataset was created by taking sampled posts from primarily bias rich sources such as Twitter and Reddit, which have indivduals saying some of the most biased and hateful speech on the internet. Furthermore, they searched within websites containing offensive languages. By choosing to draw data from these targeted sources, they were able to ensure enough examples of both hidden and obvious signs of bias to support inference tasks for the LLM. Each post was annotated using the Social Bias Frames schema, which is something that takes the categorical judgements, sich as offensiveness, and free-text descriptions of what the implied statements was trying to say to the targeted groups. These labels were given by Amazon Mechanical Turk workers, with at least three annotators per post and demographic information such as gender, race, and political leaning. While the social bias interference corpus offers insights into bias, its construction also introduces potential limitations. There is a strong focus on bias-heavy sources, meaning the dataset contains an unproportionally large number of offensive statements, which is not comparable to everyday language. This language reflects the culture on things such as Reddit and Twitter. Furthermore, whatever is being analyzed is largely subjective and shaped by the past of the person who labeled it. What this means is that certain biases may be overlooked or differently interpreted in other contexts.


In [5]:
word_counts = {} 

for post in theData.column("post"):
    words = post.lower().split()  # lowercase + split on whitespace
    for word in words:
        if word not in word_counts:
            word_counts[word] = 1
        else:
            word_counts[word]+=1

# This sorts our dictionary based off the corresponding frequency, and turns it into a list.
word_counts = sorted(word_counts.items(), key=lambda item: item[1], reverse = True)
word_counts[:20]  # Display the top 20 most frequent words

[('the', 689),
 ('a', 598),
 ('i', 481),
 ('to', 451),
 ('and', 389),
 ('of', 292),
 ('you', 288),
 ('in', 239),
 ('is', 236),
 ('that', 184),
 ('my', 177),
 ('it', 166),
 ('for', 163),
 ('.', 143),
 ('on', 130),
 ('with', 126),
 ('they', 125),
 ('have', 122),
 ('are', 121),
 ('do', 120)]

In [6]:
import string 
def count(w, r):
    total = 0
    splittedR = r.split()
    for i in splittedR:
        if i == w:
            total+=1
    return total

def total_words(r):
    """
    Returns the number of words in the Post r, separated by whitespace.

    >>> post = "hello there friend!"
    >>> total_words(post)
    3
    """
    splittedR = r.split()
    return(len(splittedR))

def df(w, lst):
    count = 0
    for doc in lst:
        docSplit = doc.split()
        for word in docSplit:
            if word == w:
                count+=1
                break
    return count

def tf_idf(w, r, lst):
    """
    Compute the tf_idf score for word w within reddit post r.
    To compute the inverse document frequency, use lst, the set of documents.
    """
    # Term Frequency (TF)
    tf = count(w, r) / total_words(r)

    # Inverse Document Frequency (IDF)
    N = len(lst)
    df_w = df(w, lst)
    idf = np.log(N /(1 + df_w))
    return tf * idf



# Convert txt files into a string
text_data = " ".join(theData.column("post")).lower().replace("&amp;", " ")
for p in string.punctuation:
    text_data = text_data.replace(p, " ")
array_text_data = text_data.split()

# Save to a text file
    

with open("beemovie.txt", "r") as file:
    bee_text = file.read()
for p in string.punctuation:
    bee_text = bee_text.replace(p, " ")

with open("constitution.txt", "r") as file:
    constitution_text = file.read()
for p in string.punctuation:
    constitution_text = constitution_text.replace(p, " ")

theData_bee_constitution = [text_data.lower(), bee_text.lower(), constitution_text.lower()]


word_highest_tf_idf = {}
word_scores = {}

# Iterate through each unique word in the pizza dataset
for word in tqdm(set(array_text_data), desc="Processing TF-IDF scores: "):
    score = tf_idf(word, text_data, theData_bee_constitution) # TF-IDF for that word, in the entire Reddit dataset, using the 3-document corpus
    word_scores[word] = score # Store the TF-IDF for that word

    max_score = max(word_scores.values())

    #This does not work!
    # for word, score in word_scores.items():
    #     if score == max_score:
    #         if word not in word_highest_tf_idf:
    #             word_highest_tf_idf[word] = 1
    #         else:
    #             word_highest_tf_idf[word] += 1

word_highest_tf_idf = sorted(word_scores.items(), key=lambda item: item[1], reverse=True)[:20]
word_highest_tf_idf


Processing TF-IDF scores: 100%|██████████| 4920/4920 [00:25<00:00, 192.45it/s]


[('fucking', 0.00093337359284506496),
 ('co', 0.00055240477943891602),
 ('https', 0.00053335633876860856),
 ('rt', 0.00053335633876860856),
 ('128514', 0.00051430789809830122),
 ('bitches', 0.00047621101675768626),
 ('ass', 0.00047621101675768626),
 ('fuck', 0.00045716257608737881),
 ('bitch', 0.00045716257608737881),
 ('shit', 0.00036192037273584154),
 ('u', 0.00034287193206553409),
 ('child', 0.00032382349139522669),
 ('hoes', 0.00030477505072491919),
 ('sex', 0.00030477505072491919),
 ('128557', 0.00030477505072491919),
 ('niggas', 0.00028572661005461179),
 ('jews', 0.00028572661005461179),
 ('releasethememo', 0.00028572661005461179),
 ('fucked', 0.00026667816938430428),
 ('kids', 0.00026667816938430428)]

The TF-IDF results tell us that the words that are frequent in our hatespeech dataset but rare in *The Bee Movie* and the US Constitution, showing us the language and topics that are unique to our dataset. The fact that words such as "fuck" and "ass" are ranked so high reflects how informal and agressive the tone used by people on online discussions can be. The high ranking of various derogatory terms indicates that many discussions revolve around targeted insults, slurs, and sexually explicit language. The less frequent terms ("releasethememo", emoji ASCIs) appear in this ranking because they are super rare when comparing across the corpus. This gives them high TF-IDF scores despite lower overall use throughout the dataset. Overall, we feel as though these results show us that sites such as Reddit, Twitter, and hate-focused forums are very casual in nature, modern in language, and house occasionally very hostile conversations that are not very similar to the neutral and formal text found in the other two documents.

# Part 2

In [7]:
to_label = theData.sample(50, with_replacement = False).select("post")
to_label.to_csv('labels.csv')
bashar = Table.read_table("labels_bashar.csv")
nishanth = Table.read_table("labels_nishanth.csv")
to_label = to_label.with_column("Bashar", bashar.column("Offensive?")).with_column("Nishanth", nishanth.column("Offensive?"))
pivoted = to_label.pivot("Bashar", "Nishanth").relabel("Nishanth","")
pivoted.show()

In [8]:
def cohen_kappa_yes_no_maybe(mm, nm, ym,
                             mn, nn, yn,
                             my, ny, yy):

    total = yy + yn + ym + ny + nn + nm + my + mn + mm

    # observed agreement
    po = (yy + nn + mm) / total

    # row marginals first rater
    row_yes   = (yy + yn + ym) / total
    row_no    = (ny + nn + nm) / total
    row_maybe = (my + mn + mm) / total

    # column marginals second rater
    col_yes   = (yy + ny + my) / total
    col_no    = (yn + nn + mn) / total
    col_maybe = (ym + nm + mm) / total

    # expected agreement
    pe = (row_yes * col_yes) + (row_no * col_no) + (row_maybe * col_maybe)

    kappa = (po - pe) / (1 - pe)
    return kappa

kappa = cohen_kappa_yes_no_maybe(pivoted[1][0], pivoted[1][1], pivoted[1][2],
                              pivoted[2][0], pivoted[2][1], pivoted[2][2],
                              pivoted[3][0], pivoted[3][1], pivoted[3][2])

kappa

0.54367666232073009

**Reflection**:

Our experience on hand coding was not at all bad, and it was pretty simple to label our data even with options for our one variable. We thought it would be tedious and challenging because there would be so much data to label in an subjective manner, but in the end, the screening/labeling process, at least for only one variable, did not take long at all. We did disagree on quite a few labels, particularly when one person labeled "maybe" and the other labeled either "yes" or "no", but we agreed on a lot more. A Kappa of ~0.54 indicates that we agreed more than would be expected by chance. 36 of our data points were the same, with agreement on 17 yes's, 18 no's, and 1 maybe. We feel as though these numbers are not out of the ordinary. One difficulty was implementing the cohen's kappa function for a 3x3 confusion matrix, which was a bit tedious to hard code. One insighted we gained is the neccesity to ensure all annotators perform their labeling in a standardized manner; this avoids the need for data cleaning/reformating at later stage when the data is ready to be aggregated. 

# Part 3

In [9]:
client = OpenAI(api_key = API_KEY)

In [10]:
promptv1 = [
    {
        "role": "system",
        "content": (
            "You are an annotator for a social-bias classification task. "
            "Label a post yes, no, or maybe for maybe based on offensiveness."
        )
    },
    {
        "role": "developer",
        "content": (
            "Instructions:\n"
            "- Output exactly one of: 'Yes', 'No', 'Maybe'.\n"
            "- Do not include explanations.\n"
            "- Profanity alone does not guarantee '1'. Consider whether the language targets a person or group."
            "\n\nExample:"
            "\nInput: I love this movie"
            "\nOptions: Yes / No / Maybe"
            "\nOutput: No"
        )
    },
    {
        "role": "user",
        "content": "" 
    }
]


In [11]:
promptv2 = [
        {
            "role": "system",
            "content": (
                        "You are an expert annotator for the Social Bias Frames dataset."
                        "Your task to assign a given social media post entry with one of the following labels based on whether"
                        "the post is offensive or not: '1.0' for yes, '0.0' for no, '0.5' for maybe based on the dataset's definition of offensiveness:"
                        "content that targets a group or individual with bias, prejudice, or harmful stereotypes. "
                        "Profanity alone is not necessarily offensive if not directed at a group or individual. "
                        "Sarcasm, in-group usage, and self-referential slurs should be judged in context. "
                        "Even if you are uncertain, you must pick exactly one label."
                    )
        },
        {
            "role": "developer",
            "content": (
                        "Instructions:"
                        "\n-Output solely the appropriate label '1.0', '0.0', or '0.5'."
                        "\n-Do not add explanations or any other text."
                        "\n-Treat group-targeted bias as offensive, even if it is phrased as a joke."
                        "\n-Profanity alone is not enough to be considered offensive; consider the intention and target of the post"
                        "\n-Do not make up information; do not output hallucinated information."
                        "\n\nExamples:"
                        "\nInput: What shines the most in a black person? The handcuffs"
                        "\nOption 1. 1.0"
                        "\nOption 2. 0.0"
                        "\nOption 3. 0.5"
                        "\nOutput: 1"
                        "\n\nInput: I love Christopher Nolan films!"
                        "\nOption 1. 1.0"
                        "\nOption 2. 0.0"
                        "\nOption 3. 0.5"
                        "\nOutput: 0"
                        "\nInput: Oh great, another woman driver holding up traffic"
                        "\nOption 1. 1.0"
                        "\nOption 2. 0.0"
                        "\nOption 3. 0.5"
                        "\nOutput: 1.0"
            )
        },
    {
        "role": "user",
        "content": ""
    }
    ]

    

In [12]:
promptv3 = [
        {
            "role": "system",
            "content": (
                        "You are an expert annotator for the Social Bias Frames dataset."
                        "Your task to assign a given social media post entry with one of the following labels based on whether"
                        "the post is offensive or not: '1.0' for yes, '0.0' for no, '0.5' for maybe based on the dataset's definition of offensiveness:"
                        "content that targets a group or individual with bias, prejudice, or harmful stereotypes. "
                        "Profanity alone is not necessarily offensive if not directed at a group or individual. "
                        "Sarcasm, in-group usage, and self-referential slurs should be judged in context. "
                        "Even if you are uncertain, you must pick exactly one label."
                        "If the target or intent is truely unclear and cannot be inferred but the content could be interpreted as offensive, output '0.5'."

                    )
        },
        {
            "role": "developer",
            "content": (
                        "Instructions:"
                        "\n-Output solely the appropriate label '1.0', '0.0', or '0.5'."
                        "\n-Do not add explanations or any other text."
                        "\n-Treat group-targeted bias as offensive, even if it is phrased as a joke."
                        "\n-Profanity alone is not enough to be considered offensive; consider the intention and target of the post"
                        "\n-Do not make up information; do not output hallucinated information."
                        "\n\nExamples:"
                        "\nInput: What shines the most in a black person? The handcuffs"
                        "\nOption 1. 1.0"
                        "\nOption 2. 0.0"
                        "\nOption 3. 0.5"
                        "\nOutput: 1"
                        "\n\nInput: I love Christopher Nolan films!"
                        "\nOption 1. 1.0"
                        "\nOption 2. 0.0"
                        "\nOption 3. 0.5"
                        "\nOutput: 0"
                        "\nInput: Oh great, another woman driver holding up traffic"
                        "\nOption 1. 1.0"
                        "\nOption 2. 0.0"
                        "\nOption 3. 0.5"
                        "\nOutput: 1.0"
                        "\nInput: Woah, are you two sisters? You look so much alike!"
                        "\nOption 1. 1.0"
                        "\nOption 2. 0.0"
                        "\nOption 3. 0.5"
                        "\nOutput: 0.5"
            )
        },
    {
        "role": "user",
        "content": ""
    }
    ]

    

        V1 was our baseline prompt to see what the model would output without many instructions. The most prominent problem with v1 of our prompt was that is lacked context about the dataset. It had no definition of what is considered "offensive", leaving the model to apply its own arbitrary criteria. Another major failure of this prompt was not properly formatting the llm outputs; the gold labels in the dataset are represented by float values (1.0, 0.0, and 0.5), while this prompt formatted the llm outputs as "yes", "no", and "maybe". This, in combination with a lack of further constraints, resulted in explanatory outputs like "Yes because..." or quantifiers such as "likely yes". This made resulted in unstandardized outputs that did not map to the gold labels, making them difficult to analyze using an f1 score.
       In v2, we followed the prompting guidelines contained in Ziem's paper closely, clearly defining the constraints for determining offensiveness; this included context regarding concepts listed in the original paper such as sarcasm and self-referential slurs. In this version of the prompt, we made it explicitly clear that the output should be strictly one of the three possible labels. This removed the case in which the model outputted something other than expected. We also included three few-shot examples in this prompt, two with offensive posts and one without, giving the model more clear reference for how to handle various cases. We implemented stricter guidelines on output, properly formatting them as strings of floats, and what is considered "offensive" in the context of the data. One major improvement was making it clear that the use of profanity does not immediately indicate whether a post is offensive or not. This was a large problem in v1, as the model often labeled neutral posts as "1.0" purely because of the fact that they used profane language. Despite including instructions to not provide explanation, the model's output was still too verbose in a few cases. It also underused the 0.5 or "maybe" label, indicating that we needed more clear-cut boundaries on what to do in the case of uncertainty.
        After calculating the performance of prompt v2, we noticed that the model never correctly labeled 0.5 cases, and barely used the label in general. In v3, we focused on improving the model’s ability to label dynamically; that is, straying away from binary outputs of either 1.0 or 0.0 and recognizing when a post's offensiveness is not as discrete. To do so, we attempted to include following line: "If the target or intent is unclear, or the content is borderline/ambiguous, output '0.5'." However, we noticed that this actually decreased model performance, as the models began to overuse the '0.5' label, and rarely did so correctly. We found middle ground between these two instances by imposing stricter requirements for a model to be allowed to use the maybe label. We also included a quick sanitization line, telling the model not to hallucinate information to fill in missing details or make unsupported assumptions. These changes aimed to encourage the model to better handle edge cases, reduce false certainty, and improve balanced use of all three labels.

In [13]:
def chat_completion(model, message):
    """
    Makes an API request to a LLM and returns the message.

    Inputs:
    model: A String representing the model we're using
    message: A list of dictionaries, where each dictionary has two entries.
    The first entry in each dictionary should have the key "role", and a value representing the role.
    The second entry in each dictionary should have the key "content", and a value representing the content.

    Returns:
    A String representing the result of an API request to the model, parsed to just get the message back.

    >>> model_choice = "gpt-4o-mini"
    >>> message = [{"role": "user", "content": "Hi there!"}]
    >>> isinstance(prompt(model_choice, message), str)
    True
    """
    completion = client.chat.completions.create(
      model=model, # Picking model 
      messages=message # Supplying the message
    )
    response_json = completion.model_dump_json() 
    response_dict = json.loads(response_json)
    return response_dict['choices'][0]['message']['content']
    
def llm_label(model, prompt, data):
    labels = []
    for post in data.column("post"):
        prompt[2]["content"] = post
        labels.append(chat_completion(model, prompt))
    return data.with_column(model, labels)



theData = llm_label("gpt-4.1", promptv3, theData)
theData = llm_label("gpt-3.5-turbo", promptv3, theData)



In [16]:
from sklearn.metrics import precision_score, recall_score, f1_score

# Sample ground truth and predicted labels
y_true = theData.column("offensiveYN").astype(str)
y_pred_41 = theData.column("gpt-4.1")
y_pred_turbo = theData.column("gpt-3.5-turbo")

labels = ["1.0", "0.0", "0.5"]

# Compute per-class precision, recall, and F1 using sklearn
precision_41 = precision_score(y_true, y_pred_41, labels=labels, average=None)
precision_turbo = precision_score(y_true, y_pred_turbo, labels=labels, average=None)

f1_per_class_41 = f1_score(y_true, y_pred_41, labels=labels, average=None, zero_division=0)
f1_per_class_turbo = f1_score(y_true, y_pred_turbo, labels=labels, average=None, zero_division=0)

macro_f1_41 = np.mean(f1_per_class_41)
macro_f1_turbo = np.mean(f1_per_class_turbo)


recall_41 = recall_score(y_true, y_pred_41, labels=labels, average=None)
recall_turbo = recall_score(y_true, y_pred_turbo, labels=labels, average=None)
print(f"Stats for 4.1: recall: {recall_41}, precision: {precision_41}, macro_f1: {macro_f1_41}")
print(f"Stats for turbo: recall: {recall_turbo}, precision: {precision_turbo}, macro_f1: {macro_f1_turbo}")
 


model_difference = theData.where("gpt-4.1", are.not_equal_to, theData.column("gpt-3.5-turbo")).select("offensiveYN", "post", "gpt-4.1", "gpt-3.5-turbo")

print(f"The models disagreed {model_difference.num_rows} times out of {theData.num_rows} labels.")
print("Here are some examples of disagreement:")
i = 0
for row in model_difference.rows:
    print(f"\nGold:{row[0]} | gpt-4.1:{row[2]} | gpt-3.5-turbo:{row[3]} | post: {row[1]}\n\n")
    i += 1
    if i == 20:
        break

Stats for 4.1: recall: [ 0.77301927  0.82038835  0.24210526], precision: [ 0.82798165  0.74778761  0.20535714], macro_f1: 0.6013955539149337
Stats for turbo: recall: [ 0.91220557  0.54126214  0.03157895], precision: [ 0.65740741  0.83208955  0.03571429], macro_f1: 0.4845091555173063
The models disagreed 328 times out of 1000 labels.
Here are some examples of disagreement:

Gold:0.0 | gpt-4.1:0.0 | gpt-3.5-turbo:1.0 | post: tomorrow, 3 fucking PM EST yall are paying for my fallen cheek surgery and heal the petals from hanahaki !



Gold:1.0 | gpt-4.1:0.0 | gpt-3.5-turbo:1.0 | post: Never met the bitch but I fucked her like I missed her



Gold:0.0 | gpt-4.1:0.0 | gpt-3.5-turbo:0.5 | post: This is a problem that affects everyone, but we have seen far more harassment targeting women.



Gold:0.5 | gpt-4.1:1.0 | gpt-3.5-turbo:0.0 | post: People think it's weird why I don't play with baby's But I was always told as a kid not to play with my food.



Gold:1.0 | gpt-4.1:0.5 | gpt-3.5-turbo:1.

__Find all the instances where the models above labeled the response
differently. Are there any patterns? Provide your own evaluation of which
model's output you agree with most and explain why. Substantiate your
evaluation with at least one example.__

As indicated by our analysis above, the models disagreed 328 times out 1000. There are a few patterns here:

3.5-turbo seems particularly sensitive to profanity:

Gold 0.0 | 4.1 0.0 | Turbo 1.0 → “THE FUCKING HARDY BOYZ AT WRESTLEMANIA…”

Gold 0.0 | 4.1 0.0 | Turbo 1.0 → “tomorrow, 3 fucking PM EST…”

Neither of these posts are offensive, yet turbo alone flags them as such. This is interesting as we explicitly prompted both models not to use profanity alone as an indicator for offensiveness. Nonetheless, this behavior is corroborated by our stats, with turbo having a high recall for "yes" labels but lower precision; this indicates that it over-labels the "offensive" category.

Turbo also seems to struggle with more nuanced cases of bias:

Gold 1.0 | 4.1 1.0 | Turbo 0.5 → “What would be the title of Jumanji if it had an entire black cast Zoomanji”

Gold 1.0 | 4.1 1.0 | Turbo 0.5 → “The only black I love is my black coffee”

Gold 1.0 | 4.1 1.0 | Turbo 0.5 → “#notsexist, but… a woman’s ‘5 minutes’…”

4.1 is often able to correctly label these kinds of posts as offensive, while turbo doesn't seem to be able to pick up on the cues, and assigns "maybe".

Both models seem to struggle in cases of uncertainty, often going different directions: 

Gold 0.5 | 4.1 0.0 | Turbo 1.0 — “SARCASMS the ability to insult idiots…”

Gold:0.5 | gpt-4.1:0.0 | gpt-3.5-turbo:1.0 | post: If you can't except the fact i drink. I'll pop one open...

Gold:0.5 | gpt-4.1:1.0 | gpt-3.5-turbo:0.0 | post: People think it's weird why I don't play with baby's But I was always told as a kid not to play with my food.

This is the most expected pattern here; in cases in which even human annotators cannot decipher a post's intention, we likely cannot expect an LLM to perform accordingly.

Overall, we agree with gpt-4.1's output more. It adhered to more strictly to the guideline that profanity does not automatically indicate offensiveness, and was more often able to pick up on nuanced cases of social bias. We think the following is a great example that demonstrates the edge 4.1 has over turbo, being able to read between the lines and correctly label a more complex example of social bias:

“What would be the title of Jumanji if it had an entire black cast Zoomanji”
Gold: 1.0 · 4.1: 1.0 __Good job 4.1!__· Turbo: 0.5